satellite/satellitedb/overlaycache.go

```diff
@@ -33,241 +33,269 @@ type overlaycache struct {
+	var conds conditions
+	if isNewNodeQuery {
+		conds.add(
+			"total_audit_count < ?",
+			criteria.AuditCount,
+		)
+	} else {
+		conds.add(
+			"total_audit_count >= ? AND total_uptime_count >= ?",
+			criteria.AuditCount, criteria.UptimeCount,
+		)
+	}
+	conds.add(`disqualified is null
+			AND suspended is null
+			AND exit_initiated_at is null`)
 
-	safeQuery := `
-		WHERE disqualified IS NULL
-		AND suspended IS NULL
-		AND exit_initiated_at IS NULL
-		AND type = ?
-		AND free_disk >= ?
-		AND (total_audit_count < ? OR total_uptime_count < ?)
-		AND last_contact_success > ?`
-	args := append(make([]interface{}, 0, 10),
-		nodeType, criteria.FreeDisk, criteria.AuditCount, criteria.UptimeCount, time.Now().Add(-criteria.OnlineWindow))
+	conds.add("type = ?", int(pb.NodeType_STORAGE))
+	conds.add("free_disk >= ?", criteria.FreeDisk)
+	conds.add("last_contact_success > ?", time.Now().Add(-criteria.OnlineWindow))
 
 	if criteria.MinimumVersion != "" {
 		v, err := version.NewSemVer(criteria.MinimumVersion)
 		if err != nil {
-			return nil, Error.New("invalid node selection criteria version: %v", err)
+			return condition{}, Error.New("invalid node selection criteria version: %v", err)
 		}
-		safeQuery += `
-			AND (major > ? OR (major = ? AND (minor > ? OR (minor = ? AND patch >= ?))))
-			AND release`
-		args = append(args, v.Major, v.Major, v.Minor, v.Minor, v.Patch)
+		conds.add(
+			"(major > ? OR (major = ? AND (minor > ? OR (minor = ? AND patch >= ?)))) AND release ",
+			v.Major, v.Major, v.Minor, v.Minor, v.Patch,
+		)
 	}
+	cond := combine(conds...)
 
-	if !criteria.DistinctIP {
-		nodes, err = cache.queryNodes(ctx, criteria.ExcludedIDs, count, safeQuery, args...)
-		if err != nil {
-			return nil, err
+	if len(criteria.ExcludedIDs) > 0 {
+		excludedIDs := " AND id NOT IN (?" + strings.Repeat(", ?", len(criteria.ExcludedIDs)-1) + ") "
+		cond.addQuery(excludedIDs)
+		for _, id := range criteria.ExcludedIDs {
+			cond.addArg(id)
 		}
-		return nodes, nil
 	}
-
-	for i := 0; i < 3; i++ {
-		moreNodes, err := cache.queryNodesDistinct(ctx, criteria.ExcludedIDs, criteria.ExcludedNetworks, count-len(nodes), safeQuery, criteria.DistinctIP, args...)
-		if err != nil {
-			return nil, err
-		}
-		for _, n := range moreNodes {
-			nodes = append(nodes, n)
-			criteria.ExcludedIDs = append(criteria.ExcludedIDs, n.ID)
-			criteria.ExcludedNetworks = append(criteria.ExcludedNetworks, n.LastNet)
-		}
-		if len(nodes) == count {
-			break
+	if criteria.DistinctIP {
+		if len(criteria.ExcludedNetworks) > 0 {
+			excludedNetworks := " AND last_net NOT IN (?" + strings.Repeat(", ?", len(criteria.ExcludedIDs)-1) + ") "
+			cond.addQuery(excludedNetworks)
+			for _, subnet := range criteria.ExcludedNetworks {
+				cond.addArg(subnet)
+			}
 		}
+		cond.addQuery(
+			`AND last_net <> ''
+			ORDER BY last_net, RANDOM()
+			) filteredcandidates`)
```

Comment:
`ORDER BY last_net, RANDOM() ) filteredcandidates`, this isn't really a condition